[KEP-4188] New Kubelet gRPC API returning node-local Pod info

briansonnenberg opened the pull request:
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

This PR introduces a new gRPC endpoint in the Kubelet for querying information about pods running on the node. This is part of the implementation of KEP-4188. The new API will allow other components to get pod information directly from the Kubelet.

#### Which issue(s) this PR is related to:

KEP: https://github.com/kubernetes/enhancements/issues/4188

#### Special notes for your reviewer:

N/A

#### Does this PR introduce a user-facing change?

```release-note
A new gRPC service is added to the Kubelet that provides information about pods running on the node.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/4188-kubelet-pod-readiness-api
- [Example client]: https://github.com/briansonnenberg/kubelet_pods_reader/blob/main/main.go
```

briansonnenberg committed:
[KEP-4188] New Kubelet gRPC API with endpoint returning local Pod information

matthyx commented:
/cc

matthyx commented:
/ok-to-test

liggitt commented on staging/src/k8s.io/kubelet/pkg/apis/pods/v1alpha1/api.proto:
```diff
@@ -0,0 +1,57 @@
+
+// GetPodResponse returns a Pod.
+message GetPodResponse {
+    k8s.io.api.core.v1.Pod pod = 1;
```
We should not reference Kubernetes REST API types from protoc-generated GRPC APIs.

The Kubernetes API types are *not* generated as standard protoc messages ... mixing them into uses by grpc / protoc generated APIs is not something we should do. They are treated as "aberrant messages" in a best-effort way by the standard proto libraries.

xref https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/5589-gogo-dependency#motivation for details

briansonnenberg replied:
Sure, so you're suggesting we redefine Pod and all of its subfields, and there are no guarantees this will be kept in sync with other definitions of Pod?

liggitt commented on pkg/kubelet/apis/pods/server.go:
```diff
@@ -0,0 +1,437 @@
+		}
+
+		// Find the field in the source struct.
+		// First, try to match by JSON tag. If that fails, fall back to Go field name.
```
fieldmasks become part of the API surface ... I'm somewhat surprised to see an ad hoc implementation of it against REST API types here ... are we sure this works reliably across things like custom json marshaling (for metav1.Time, Quantity, etc) and inlined fields?

briansonnenberg replied:
We have unit tests for Time.  Which inlined fields were you thinking of?  Only mention of inlining I see in the Pod spec is ephemeral containers, which I added a unit test for.

ajaysundark commented:
/cc

matthyx commented on test/e2e_node/pods_api_test.go:
```diff
@@ -0,0 +1,258 @@
+			if conn != nil {
+				conn.Close()
+			}
+			ginkgo.By("Emitting Shutdown false signal; cancelling the shutdown")
```
I am not sure we need this false shutdown signal

matthyx replied:
resolved

matthyx commented on test/e2e_node/pods_api_test.go:
```diff
@@ -0,0 +1,258 @@
+					return "", err
+				}
+				return podFromDeletedEvent.ObjectMeta.UID, nil
+			}, "1m", "1s").Should(gomega.Equal(testPod.UID), "did not receive DELETED event for the test pod")
```
this line is flaky

matthyx replied:
resolved

matthyx reviewed (commented):
A few comments, overall very good, I didn't expect to be so much work but you did great!
Just update gofmt, eventually the linter check and we can assign @SergeyKanzhelev

matthyx commented:
CI failures likely linked to https://github.com/kubernetes/kubernetes/issues/134998

SergeyKanzhelev commented on pkg/kubelet/apis/pods/server.go:
```diff
@@ -0,0 +1,508 @@
+/*
+Copyright 2020 The Kubernetes Authors.
```
just remove all years from headers. Now header may omit the year and I do not need to comment on the wrong year =)

SergeyKanzhelev commented on pkg/kubelet/apis/pods/server.go:
```diff
@@ -0,0 +1,508 @@
+	if len(fieldMasks) == 0 {
+		return nil, nil
+	}
+	if len(fieldMasks) > 1 {
```
why? It is comma separated list, why not allow to split it into multiple headers. At least in Http it is a common practice and semantically interchangeable - list in one header comma separated vs. multiple headers. Furthermore, for http some client libraries are opinionated on what to do with multi-value comma separated headers

SergeyKanzhelev commented on pkg/kubelet/apis/pods/server.go:
```diff
@@ -0,0 +1,508 @@
+
+// applyFieldMask uses Go's standard reflection to copy fields from src to dest based on the mask.
+// It supports traversing into slices to select specific fields from the elements.
+func ApplyFieldMask(mask *fieldmaskpb.FieldMask, src, dest interface{}) error {
```
this looks expensive. Is any of it cache-able? Also individual field masks were not validated. And instead of invalid argument user will get simlpe erroras a return if they make a syntax error

SergeyKanzhelev commented on pkg/kubelet/apis/pods/server.go:
```diff
@@ -0,0 +1,508 @@
+
+const (
+	// FieldMaskMetadataKey is the key for the fieldmask in the gRPC metadata.
+	FieldMaskMetadataKey = "x-kubernetes-fieldmask"
```
reading the KEP I was under impression it is something more standard. I guess reflection is standard enough to guarantee back compat and versioning is not needed on this field. But we need some sort of spec here

SergeyKanzhelev commented on pkg/kubelet/server/server.go:
```diff
@@ -258,6 +261,32 @@ func ListenAndServePodResources(ctx context.Context, endpoint string, providers
+	go func() {
+		if err := server.Serve(l); err != nil && !errors.Is(err, grpc.ErrServerStopped) {
+			logger.Error(err, "Failed to serve")
+			os.Exit(1)
```
do we exit kubelet here completely? Sounds harsh.

For device plugin we are doing health monitoring. This bug fix may give a pointer: https://github.com/kubernetes/kubernetes/pull/135153

I also think we delete-recreate a folder on kubelet restart. This may be useful here as well. We had good and bad from this behavior (removing the whole device plugin directory on restart). But the good thing is that some issues are remediated by the recreation of that directory.

cc @ffromani  again

briansonnenberg replied:
Fair enough.  This logic was shamelessly copied from PodResources as well, so we can patch both with whatever fix we decide on.

SergeyKanzhelev commented on pkg/kubelet/kubelet.go:
```diff
@@ -691,7 +694,12 @@ func NewMainKubelet(ctx context.Context,
+	broadcaster := pods.NewBroadcaster()
+	klet.podsServer = pods.NewPodsServer(broadcaster)
+	podSubscribers := []subscription.PodUpdateSubscriber{klet.podsServer}
+	statusSubscribers := []subscription.StatusUpdateSubscriber{klet.podsServer}
```
Can we hide this under the feature gate?

briansonnenberg replied:
Done.

SergeyKanzhelev reviewed (approved):
/lgtm
/approve

SergeyKanzhelev commented:
@dims or @liggitt around for hack approve?

dims commented:
/approve

briansonnenberg commented:
Had to run go fmt

@SergeyKanzhelev @dims

SergeyKanzhelev commented:
/lgtm

SergeyKanzhelev commented:
need API approver now.

/assign @liggitt

liggitt commented:
/unassign

liggitt reviewed (commented):
(I did a 5 minute sweep of the proto file just to get you started on comments other API reviewers are likely to leave as well)

liggitt commented on staging/src/k8s.io/kubelet/pkg/apis/pods/v1alpha1/api.proto:
A general comment I've made during the design discussions is that the kubelet has multiple views of pods:
1. latest state observed from the API server? If so, I thought part of the motivation was to have node-local observers of the kubelet's view of pods which could keep coasting even if the connection to the API server was disrupted...
2. latest state calculated from the container runtime? If so, I'd be interested to know how the delete watch event is handled when the API server thinks a pod is deleted, and the container runtime is still running the pod.
3. maybe others...? (@tallclair and @smarterclayton would probably know)

It is not clear which of these states this API is returning. At the very least we need to be super clear about which this is returning, or make the client indicate which view they want

briansonnenberg commented on staging/src/k8s.io/kubelet/pkg/apis/pods/v1alpha1/api.proto:
```diff
@@ -0,0 +1,55 @@
+}
+
+//
+// FieldMask can be specified for any request via gRPC metadata
```
Even if we end up with edge cases or performance implications, the fixes for those are server implementation details that can be fixed without changing the API surface—that's what Alphas are for.

It makes too much sense to have masking for these large pod objects to not have it for something like this.

briansonnenberg replied:
We can remove it for now, but I'm not sure it's necessary given that it's entirely opt-in from the Client.

SergeyKanzhelev commented:
> that was mainly about reservations in the approach due to concerns over

@deads2k just to clarify, the concerns (and what you think should be a north star) are going all the way to client-libraries compatible REST API instead of gRPC, not just types? Is it a correct understanding of a concern? I might have misunderstood the discussion we had

briansonnenberg commented:
> that was mainly about reservations in the approach due to concerns over...

To be clear, most of those concerns were already considered when we accepted the KEP and opted for the gRPC API.

1. A full API-server-level interface was decided to be too heavy for the use-case, especially when we already serve similar (PodResources) information through this lighter weight mechanism.
2. This is serving the current state of Pods from the perspective of Kubelet, so what is or is not valid from other perspectives isn't a concern.
3. We require admin permissions on the host to access the API, permissions scoping can be added down the line if needed.
4. We are starting with everything and can whittle down when permissions are added.  There is no "too much information" when a host admin is the client—that was decided during our talks with sig-sec.

jpbetz reviewed (commented):
I'm not an API approver, but have a few suggestions

matthyx commented:
> I'm not an API approver, but have a few suggestions

OMG, right, I don't know why I thought you were... maybe because of your involvement in PRRs.
```
  api-approvers:
    - deads2k
    - msau42
    - smarterclayton
    - thockin
    - liggitt
```

ffromani commented:
/triage accepted
/priority important-longterm

rphillips commented:
My concerns are the following:

* Have we figured out how much more memory this implementation uses? It duplicates the pod over again?
* We are serializing the pods to grpc; what sort of overhead does this add to the kubelet? GC?
* Pod workers is designed to work on UIDs and is the authoritative source of what is running and being terminated.
* I don't see any tests for static pods. Static pods are tracked by UID and name.
* An improved approach might be to pass the UID instead of the pod to the grpc handler and have the grpc handler look up the UID in podManager.
* The enhancement states that pods can contain secured information, most cloud solutions do not block hostPath volume mounts by default, which would expose this GRPC endpoint to any workload.

ajaysundark commented on pkg/kubelet/apis/pods/server.go:
```diff
@@ -0,0 +1,280 @@
+
+type PodWatchEvent struct {
+	Type watch.EventType
+	Pod  *v1.Pod
```
This could be updated to send only pod's UID. WatchPods could fetch `podManager.GetPodByUID()` so we work with a single consistent pod cache in Kubelet. This will also save the memory footprint as called out by earlier reviews

briansonnenberg committed:
Refactor PodsServer to use PodManager as source of truth

- Fixed version in kube_features.go after rebase (1.35->1.36)
- Removed internal pod cache in PodsServer to reduce memory footprint and avoid duplication.
- Injected pod.Manager into PodsServer to serve as the single source of truth for pod data.
- Refactored WatchPods to broadcast UIDs and fetch fresh pod data from podManager, ensuring consistency.
- Updated convertWatchEventType to safely handle unknown event types.
- Refactored unit tests to use MockManager and added a test case for static pods.
- Updated e2e suite with static pod test

briansonnenberg committed:
Fix typo

briansonnenberg commented:
/retest

briansonnenberg committed:
fmt and ci fixes

briansonnenberg committed:
linter fixes

briansonnenberg commented:
> My concerns are the following:
> 
> * Have we figured out how much more memory this implementation uses? It duplicates the pod over again?
> * We are serializing the pods to grpc; what sort of overhead does this add to the kubelet? GC?
> * Pod workers is designed to work on UIDs and is the authoritative source of what is running and being terminated.
> * I don't see any tests for static pods. Static pods are tracked by UID and name.
> * An improved approach might be to pass the UID instead of the pod to the grpc handler and have the grpc handler look up the UID in podManager.
> * The enhancement states that pods can contain secured information, most cloud solutions do not block hostPath volume mounts by default, which would expose this GRPC endpoint to any workload.

Thank you for the feedback @rphillips and @ajaysundark! I have made the changes to use the UID, as well as the tests for the static pods.

Re: serialization overhead, I ran a quick benchmark of 100 complex, heavy pods and see 28 microseconds and 3 allocations per pod, memory consumption is ~1:1 with the size of the pod object being serialized. Biggest optimization opportunity here is to not serialize per watcher, and instead serialize once and send to all watchers.

Re: hostPath, this seems more of a concern of understanding the implications of hostPath than with the use of filesystem permissions itself.  This wasn't flagged as a concern during our security review.

liggitt commented on pkg/kubelet/pod_workers.go:
```diff
@@ -984,6 +990,16 @@ func (p *podWorkers) UpdatePod(options UpdatePodOptions) {
+	for _, s := range p.subscribers {
+		switch options.UpdateType {
+		case kubetypes.SyncPodCreate:
+			s.OnPodAdded(pod)
```
how does the `pod` passed to the subscriber relate to `status.pendingUpdate.Pod` as modified in this function (e.g. on line 989 above)?

this is a good example of the multiple copies/versions of the pod the kubelet has in view, and the possible confusion over which view a subscriber to this API is getting

liggitt commented on staging/src/k8s.io/kubelet/pkg/apis/pods/v1alpha1/api.proto:
```diff
@@ -0,0 +1,57 @@
+
+service Pods {
+    // WatchPods will initially return an ADDED WatchPodsEvent for each
+    // existing Pod, and then return a WatchPodsEvent whenever a Pod is created,
```
how does the watcher know when the full set of ADDED events for initial pods has been received?

liggitt commented on pkg/kubelet/apis/pods/server.go:
```diff
@@ -0,0 +1,273 @@
+			if err != nil {
+				logger.Error(err, "Error marshalling watch event pod")
+				metrics.PodWatchEventsDroppedTotal.Inc()
+				continue
```
as a general comment, ignoring/continuing when there are errors sending a pod to a watcher is very difficult to reason about... I'd expect us to be closing the watcher when that happens so they don't end up with incomplete and inconsistent states

briansonnenberg committed:
Refactored broadcast logic, specifically:

For ADDED, we broadcast directly after receiving the pod info and successfully processing it for the first time from the API server, so this should always be the first thing broadcast for a given pod.

For DELETED, we broadcast after it has been processed in updateStatusInternal with podIsFinished, so this should always be the last thing broadcast for a given pod.

For MODIFIED, we broadcast once when we receive a new spec from the API server, and once when we finish processing in the status manager.  This way the watchers can see the flow from when a spec is received to when it has reconciled.  To avoid status flapping on slow status updates with many spec updates, we always overlay the latest status from the status manager on every broadcast.  This hybrid state is fine because the ObservedGeneration will indicate how far desynced the spec has become from the status, so watchers can act accordingly.

briansonnenberg committed:
Update generated protobuf bindings for PodsAPI v1alpha1

liggitt commented on staging/src/k8s.io/kubelet/pkg/apis/pods/v1alpha1/api.proto:
```diff
@@ -0,0 +1,58 @@
+}
+
+enum EventType {
+    ADDED = 0;
```
nit: make ADDED a non-zero value so WatchPodsEvent{} isn't automatically an ADDED event

liggitt commented on staging/src/k8s.io/kubelet/pkg/apis/pods/v1alpha1/api.proto:
```diff
@@ -0,0 +1,58 @@
+message WatchPodsEvent {
+    // ADDED, MODIFIED, DELETED, INITIAL_SYNC_COMPLETE
+    EventType type = 1;
+    bytes pod = 2;
```
pod is nil for INITIAL_SYNC_COMPLETE, right? if so, document that

smarterclayton commented on pkg/kubelet/pod_workers.go:
```diff
@@ -984,6 +990,18 @@ func (p *podWorkers) UpdatePod(options UpdatePodOptions) {
 	}
 	status.working = true
 	klog.V(4).InfoS("Notifying pod of pending update", "pod", klog.KRef(ns, name), "podUID", uid, "workType", status.WorkType())
+	switch options.UpdateType {
```
Having subscribers be synchronous while holding the pod lock is.... bad.  That means subscribers block the core notification loop of the kubelet to change pod state.

How complex can a subscriber be?  Can we move this out of being under the pod lock and into a worker lock?

briansonnenberg committed:
Addressed comments and optimized broadcaster.

Removed split-brain subscriber interface and switched to purely PodWorker broadcasting immediately after SetPodStatus to ensure consistent, monotonic view of the pod.

liggitt commented on pkg/kubelet/pod_workers.go:
```diff
@@ -1745,6 +1745,7 @@ func (p *podWorkers) requeueLastPodUpdate(podUID types.UID, status *podSyncStatu
 		return
 	}
 	copied := *status.activeUpdate
+	copied.UpdateType = kubetypes.SyncPodSync
```
was this a bugfix or a spurious change that should be dropped?

briansonnenberg committed:
Addressed comments. Change OnPodDeleted to only populate the name, namespace and UID of the pod object to avoid sending confusing, stale end state.

briansonnenberg committed:
Overlay latest status for GET and LIST, and in the initial sync for WATCH.

briansonnenberg committed:
ci fixes

briansonnenberg committed:
Moved broadcast logic into StatusManager with no lock-holding.

smarterclayton reviewed (changes requested):
Three final comments (I think)

briansonnenberg commented on pkg/kubelet/pod_workers.go:
```diff
@@ -777,7 +777,6 @@ func (p *podWorkers) UpdatePod(ctx context.Context, options UpdatePodOptions) {
 	updateLogger := logger.WithValues("pod", klog.KRef(ns, name), "podUID", uid, "updateType", options.UpdateType)
 	p.podLock.Lock()
 	defer p.podLock.Unlock()
-
```
Ack
